abopt/base.py

```diff
@@ -168,8 +168,12 @@ def complete(self, state):
         self.znorm = dot(self.z, self.z) ** 0.5
         self.xnorm = dot(self.x, self.x) ** 0.5
         self.Pxnorm = dot(self.Px, self.Px) ** 0.5
-        self.complete_y(state)
-        self.complete_g(state)
+        
+        if self.problem.problem_dual_eval:
+            self.complete_y_g(state)
```

Comment:
I see what you are concerned about. accessing _member from outside of the class.

The branching can be done inside Problem.f_g():

if self._objective_gradient:
   return ...
else:
  return self.f(), self.g()

Reply:
done. The branching is now in Problem.fg(x). In Proposal I removed complete_y and complete_g and replaced it by complete_y_g, this required replacing complete_y in the linesearch by complete_y_g as well

Reply:
is that what you meant?